File: java/src/leetcode/recursion/ReversePrint.java
package leetcode.recursion;

import java.util.ArrayList;
import java.util.List;

/**
 * @author chengzw
 * @description 从尾到头打印链表，https://leetcode-cn.com/problems/cong-wei-dao-tou-da-yin-lian-biao-lcof/
 * @since 2021/9/7
 */

/**
 * 思路：递归找链表节点的值，用全局变量添加 head.val
 * 时间复杂度：O(n)
 * 空间复杂度：O(n)
 */
public class ReversePrint {
    List<Integer> list = new ArrayList();

    public int[] reversePrint(ListNode head) {
        reversePrint_r(head);

        //ArrayList 转为数组
        int[] result = new int[list.size()];
        int i = 0;
        for (int k : list) {
            result[i++] = k;
        }
        return result;
    }

    public void reversePrint_r(ListNode head) {
        if (head == null) return;
        reversePrint_r(head.next);
        list.add(head.val);
    }
}

File: go/leetcode/recursion/reverse_print.go
package main

/**
 * @description 剑指 Offer 06. 从尾到头打印链表
 * @author chengzw
 * @since 2022/6/6
 * @link https://leetcode.cn/problems/cong-wei-dao-tou-da-yin-lian-biao-lcof/
 */

/**
输入一个链表的头节点，从尾到头反过来返回每个节点的值（用数组返回）。

示例 1：
输入：head = [1,3,2]
输出：[2,3,1]
*/

type ListNode struct {
	Val  int
	Next *ListNode
}

/**
思路：递归找链表节点的值，将 head 的 val 加上递归调用的结果上
时间复杂度：O(n)
空间复杂度：O(n)
*/

func reversePrint(head *ListNode) []int {
	if head == nil {
		return nil
	}
	return append(reversePrint(head.Next), head.Val)
}
